File: evals/behavioral-analysis/data/arbitrary-resource-read-write/typescript/directory_listing.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

/**
 * Example: Directory Listing Information Disclosure
 * THREAT: Exposes directory structure and sensitive file names
 */
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import * as fs from "fs";
import * as path from "path";

const server = new Server(
  { name: "dir-listing-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

function recursiveList(dir: string, depth: number = 0, maxDepth: number = 10): string[] {
  if (depth > maxDepth) return [];
  
  const results: string[] = [];
  try {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      results.push(fullPath);
      if (entry.isDirectory()) {
        results.push(...recursiveList(fullPath, depth + 1, maxDepth));
      }
    }
  } catch {}
  
  return results;
}

server.setRequestHandler("tools/call", async (request: any) => {
  const { name, arguments: args } = request.params;

  if (name === "list_directory") {
    const dirPath = args.path;  // Tainted input
    const recursive = args.recursive || false;
    
    // VULNERABLE: No path validation
    // User can list /etc, /home, ~/.ssh, etc.
    try {
      if (recursive) {
        const files = recursiveList(dirPath);
        return { content: [{ type: "text", text: files.join("\n") }] };
      } else {
        const entries = fs.readdirSync(dirPath);
        return { content: [{ type: "text", text: entries.join("\n") }] };
      }
    } catch (error: any) {
      return { content: [{ type: "text", text: `Error: ${error.message}` }] };
    }
  }

  if (name === "search_files") {
    const searchPath = args.searchPath;  // Tainted input
    const pattern = args.pattern;        // Tainted input
    
    // VULNERABLE: Search in arbitrary directories
    const allFiles = recursiveList(searchPath);
    const matches = allFiles.filter(f => f.includes(pattern));
    
    return { content: [{ type: "text", text: matches.join("\n") }] };
  }

  return { content: [{ type: "text", text: "Unknown tool" }] };
});

const transport = new StdioServerTransport();
server.connect(transport);

File: evals/behavioral-analysis/data/arbitrary-resource-read-write/go/directory_listing.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Example: Directory Listing
// THREAT: Lists arbitrary directories
package main

import (
	"context"
	"log"
	"os"
	"path/filepath"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type ListDirectoryInput struct {
	Path string `json:"path" jsonschema:"directory path to list"`
}

type ListDirectoryOutput struct {
	Files string `json:"files" jsonschema:"list of files"`
}

func ListDirectory(ctx context.Context, req *mcp.CallToolRequest, input ListDirectoryInput) (*mcp.CallToolResult, ListDirectoryOutput, error) {
	// ARBITRARY RESOURCE READ: List any directory
	var files []string
	filepath.Walk(input.Path, func(path string, info os.FileInfo, err error) error {
		if err == nil {
			files = append(files, path)
		}
		return nil
	})

	result := ""
	for _, f := range files {
		result += f + "\n"
	}
	return nil, ListDirectoryOutput{Files: result}, nil
}

func main() {
	server := mcp.NewServer(&mcp.Implementation{Name: "directory-listing", Version: "v1.0.0"}, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "list_directory", Description: "List directory contents"}, ListDirectory)
	if err := server.Run(context.Background(), &mcp.StdioTransport{}); err != nil {
		log.Fatal(err)
	}
}
